### ISSUE TITLE
Restarting a reactor should raise an exception

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2066]
|<img alt="TimothyFitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| TimothyFitz reported|
|-|-|
|Trac ID|trac#2066|
|Type|enhancement|
|Created|2006-09-07 19:00:24Z|
|Branch|https://github.com/twisted/twisted/tree/reactor-restart-raise-2066|

About once a month someone asks twisted-python why they have really strange behaviour, and happen to be restarting the reactor. Restarting the reactor is unsupported (currently, by all implementors, nothing prevents it from being implemented) so these all fail, but always fail poorly. A call to reactor.run() after reactor.stop() should raise an exception that clearly says "This reactor does not support restarting" or "Reactors do not support restarting". 

Attachments:

* [2066.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2df/2dfa388bde651ad68456fac372ecc71da44c11af/5277d69d68427a0c11f6deee0031a547e142aca3.patch) (2799 bytes) - added by mishok13 on 2011-03-14 02:56:36Z - First attempt at the actual fix

<details><summary>Searchable metadata</summary>

```
trac-id__2066 2066
type__enhancement enhancement
reporter__TimothyFitz TimothyFitz
priority__highest highest
milestone__ 
branch__branches_reactor_restart_raise_2066 branches/reactor-restart-raise-2066
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1157655624000000 1157655624000000
changetime__1300088956000000 1300088956000000
version__None None
owner__ 
cc__TimothyFitz
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="TimothyFitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|TimothyFitz commented|
|-|-|

Glyph Lefkowitz
>On Thu, 7 Sep 2006 15:01:08 -0400, James Y Knight <foom@...> wrote:
>>However, even though it'd be possible to fix, perhaps it is better to
>>simply throw an exception on the second reactor.run(), since most  people
>>who do that shouldn't actually be running the reactor twice,  they just
>>didn't know the better thing to do yet.
>
>I think that's a good idea.  Based on the nature of the questions that I've seen, I >think that this would probably be a good idea even if we *did* have a restartable >reactor.  Add an interface, IRestartableReactor, and require people who want to use >it to call 'reactor.restart()'.  While this might not actually *do* anything, the >conspicuous absence of it from any example code and a potentially useful error >message from reactor.run would prevent the question from being asked repeatedly...

This bug will cover making reactor.run() raise on a second call. Someone else can pick up IRestartableReactor and the like.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="mishok13's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|mishok13 commented|
|-|-|

The test is basically a rewrite of test_multipleRun and the patch introduces new instance variable (which sucks).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh commented|
|-|-|

(In [31008]) Branching to 'reactor-restart-raise-2066'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_7"></a>|@lvh commented|
|-|-|

(In [31009]) Apply patch, refs [#2066](https://github.com/twisted/twisted/issues/2066)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_7"></a>|@lvh commented|
|-|-|

Looks great! Merging.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_7"></a>|@lvh commented|
|-|-|

(Okay, that's actually a lie: it's missing a topfile, but I'm going to add that for you)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_11"></a>|@lvh set status to `closed`|
|-|-|

(In [31011]) Raise when you try to restart an unrestartable reactor

Author: mishok13
Reviewer: lvh
Fixes: [#2066](https://github.com/twisted/twisted/issues/2066)

Unrestartable reactors not complaining was confusing. When you try to do it, you get an exception now, which is a much nicer failure mode.
